core/src/main/java/org/elasticsearch/script/CachingCompiler.java

```diff
@@ -0,0 +1,385 @@
+            if (logger.isDebugEnabled()) {
+                logger.debug("removed {} from cache, reason: {}", notification.getValue(),
+                        notification.getRemovalReason());
+            }
+            scriptMetrics.onCacheEviction();
+        }
+    }
+
+    private class DirectoryChangesListener implements FileChangesListener {
+        private Tuple<String, String> getScriptNameExt(Path file) {
+            Path scriptPath = scriptsDirectory.relativize(file);
+            int extIndex = scriptPath.toString().lastIndexOf('.');
+            if (extIndex <= 0) {
+                return null;
+            }
+
+            String ext = scriptPath.toString().substring(extIndex + 1);
+            if (ext.isEmpty()) {
+                return null;
+            }
+
+            String scriptName = scriptPath.toString().substring(0, extIndex)
+                    .replace(scriptPath.getFileSystem().getSeparator(), "_");
+            return new Tuple<>(scriptName, ext);
+        }
+
+        @Override
+        public void onFileInit(Path file) {
+            Tuple<String, String> scriptNameExt = getScriptNameExt(file);
+            if (scriptNameExt == null) {
+                logger.debug("Skipped script with invalid extension : [{}]", file);
+                return;
+            }
+            if (logger.isTraceEnabled()) {
+                logger.trace("Loading script file : [{}]", file);
+            }
+
+            CacheKeyT cacheKey = cacheKeyForFile(scriptNameExt.v1(), scriptNameExt.v2());
+            if (cacheKey == null) {
+                return;
+            }
+            try {
+                /* we don't know yet what the script will be used for, but if all of the operations
+                 * for this lang with file scripts are disabled, it makes no sense to even compile
+                 * it and cache it. */
+                if (areAnyScriptContextsEnabled(cacheKey, ScriptType.FILE)) {
+                    logger.info("compiling script file [{}]", file.toAbsolutePath());
+                    try (InputStreamReader reader = new InputStreamReader(
+                            Files.newInputStream(file), StandardCharsets.UTF_8)) {
+                        String body = Streams.copyToString(reader);
+                        fileScripts.put(cacheKey, compileFileScript(cacheKey, body, file));
+                        scriptMetrics.onCompilation();
+                    }
+                } else {
+                    logger.warn("skipping compile of script file [{}] as all scripted operations "
+                            + "are disabled for file scripts", file.toAbsolutePath());
+                }
+            } catch (ScriptException e) {
+                /* Attempt to extract a concise error message using the xcontent generation
+                 * mechanisms and log that. */
+                try (XContentBuilder builder = JsonXContent.contentBuilder()) {
+                    builder.prettyPrint();
```

Comment:
This is in response to seeing a file. The idea is that we actually have pretty good error messages in the pretty dump of the exception so we can reuse them for file scripts.